libraries/chain_kv/include/b1/chain_kv/chain_kv.hpp

```diff
@@ -110,12 +111,141 @@ inline bytes get_next_prefix(const bytes& prefix) {
+      constexpr std::size_t size() const { return sizeof(T); }
+   };
+
+   template<typename T>
+   struct value_storage {
+      char* as_char_ptr(T& value) const { return reinterpret_cast<char*>(&value); }
+      constexpr std::size_t size() const { return sizeof(T); }
+   };
+
+   template <typename T, std::size_t N>
+   auto append_key(bytes& dest, T value) {
+      constexpr auto type_size = value_storage<T>().size();
+      constexpr auto buf_size = type_size * N;
+      char buf[buf_size];
+      detail::copy_and_swap_to_buffer<type_size, N>(buf, value_storage<T>().as_char_ptr(value));
+      std::reverse(std::begin(buf), std::end(buf));
+      dest.insert(dest.end(), std::begin(buf), std::end(buf));
+   }
+
+   template <typename UInt, typename T>
+   UInt float_to_key(T value) {
+      static_assert(sizeof(T) == sizeof(UInt), "Expected unsigned int of the same size");
+      UInt result;
+      std::memcpy(&result, &value, sizeof(T));
+      const UInt signbit = (static_cast<UInt>(1) << (std::numeric_limits<UInt>::digits - 1));
+      UInt mask    = 0;
+      if (result == signbit)
+         result = 0;
+      if (result & signbit)
+         mask = ~mask;
+      return result ^ (mask | signbit);
+   }
+
+   template <typename T, typename UInt>
+   T key_to_float(UInt value) {
+      static_assert(sizeof(T) == sizeof(UInt), "Expected unsigned int of the same size");
+      // encoded signbit indicates positive value
+      const UInt signbit = (static_cast<UInt>(1) << (std::numeric_limits<UInt>::digits - 1));
+      UInt mask    = 0;
+      if ((value & signbit) == 0)
+         mask = ~mask;
+      value = {value ^ (mask | signbit)};
+      T float_result;
+      std::memcpy(&float_result, &value, sizeof(UInt));
+      return float_result;
+   }
+
+   template<typename Key, std::size_t N>
+   bool extract_key(bytes::const_iterator& key_loc, bytes::const_iterator key_end, Key& key) {
+      const auto distance = std::distance(key_loc, key_end);
+      constexpr auto type_size = value_storage<Key>().size();
+      constexpr auto key_size = type_size * N;
+      if (distance < key_size)
+         return false;
+
+      key_end = key_loc + key_size;
+      bytes key_store(key_loc, key_end);
+      key_loc = key_end;
+      std::reverse(std::begin(key_store), std::end(key_store));
+      detail::copy_and_swap_to_buffer<type_size, N>(value_storage<Key>().as_char_ptr(key), key_store.data());
+
+      return true;
+   }
+}
+
 template <typename T>
 auto append_key(bytes& dest, T value) -> std::enable_if_t<std::is_unsigned_v<T>, void> {
-   char buf[sizeof(value)];
-   memcpy(buf, &value, sizeof(value));
-   std::reverse(std::begin(buf), std::end(buf));
-   dest.insert(dest.end(), std::begin(buf), std::end(buf));
+   detail::append_key<T, 1>(dest, value);
+}
+
+template <typename T, std::size_t N>
+auto append_key(bytes& dest, std::array<T, N> value) -> std::enable_if_t<std::is_unsigned_v<T>, void> {
```

Comment:
Is this style acceptable in eosio?  I hope you say it is because I strongly prefer it.

Reply:
Not sure which part of the style you are referring to, but if follows the style of the append_key method that was already there (or at least as far as I notice)

Reply:
This form of function declaration is required here because the return type is referring to the template parameter T. I have no preference, but I doubt the majority would prefer using the arrow syntax in every function.

Reply:
The return type can always name template parameters.  The trailing return type allows referencing `dest` and `value` as well, but that isn't being used here.